File: backend/find_mixed_queries.js

const fs = require('fs');
const path = require('path');

function walk(dir) {
    let results = [];
    const list = fs.readdirSync(dir);
    list.forEach(file => {
        file = path.join(dir, file);
        const stat = fs.statSync(file);
        if (stat && stat.isDirectory()) {
            if (!file.includes('node_modules')) {
                results = results.concat(walk(file));
            }
        } else {
            if (file.endsWith('.js')) {
                results.push(file);
            }
        }
    });
    return results;
}

const files = walk('./backend');

files.forEach(file => {
    const content = fs.readFileSync(file, 'utf8');
    // Search for prisma calls that have both select: and include: in the same object block
    // This is a naive regex but might catch simple cases
    const regex = /prisma\.[a-zA-Z]+\.[a-zA-Z]+\(\{[\s\S]*?\}\)/g;
    let match;
    while ((match = regex.exec(content)) !== null) {
        const block = match[0];
        if (block.includes('select:') && block.includes('include:')) {
            console.log(`Found mixed query in ${file}:`);
            console.log(block);
            console.log('-------------------');
        }
    }
});

File: backend/find_mixed_queries.py

import os
import re

def find_mixed_queries(root_dir):
    for root, dirs, files in os.walk(root_dir):
        if 'node_modules' in dirs:
            dirs.remove('node_modules')
        for file in files:
            if file.endswith('.js'):
                path = os.path.join(root, file)
                with open(path, 'r', encoding='utf-8') as f:
                    content = f.read()
                    # Find prisma calls using a regex that captures the call block
                    # prisma.model.method({...})
                    pattern = r'prisma\.[a-zA-Z]+\.[a-zA-Z]+\(\{[\s\S]*?\}\)'
                    for match in re.finditer(pattern, content):
                        block = match.group(0)
                        if 'select:' in block and 'include:' in block:
                            print(f"Found mixed query in {path}:")
                            print(block)
                            print("-" * 20)

if __name__ == "__main__":
    find_mixed_queries('backend')
